### PR TITLE
Add dirty flag in resolution

### PR DESCRIPTION
**Pull request by FFY00**
Solves https://github.com/libratbag/libratbag/issues/436

I was only able to check in my Steelsereis Rival 310. Please test on other mice.

### CODE REVIEW

File: src/driver-hidpp20.c [hunk 1]
```diff
@@ -1166,6 +1166,9 @@ hidpp20drv_commit(struct ratbag_device *device)
 		drv_data->profiles->profiles[profile->index].enabled = profile->is_enabled;
 
 		ratbag_profile_for_each_resolution(profile, resolution) {
+			if (!resolution->dirty)
+				continue;
+
```
Comment (bentiss) [inline src/driver-hidpp20.c]: This test needs to be placed after `dpi_index = resolution->index;` or changing any other setting will switch the resolution to the first one (`dpi_index` will stay at `0`)

File: src/libratbag-private.h [hunk 2]
```diff
@@ -267,6 +267,7 @@ struct ratbag_resolution {
 
 	bool is_active;
 	bool is_default;
+	bool dirty;			/**< profile changed since last commit */
```
Comment (bentiss) [inline src/libratbag-private.h]: `s/profile/resolution/`

File: src/driver-etekcity.c [hunk 3]
```diff
@@ -457,6 +457,9 @@ etekcity_read_profile(struct ratbag_profile *profile)
 	}
 
 	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (!resolution->dirty)
+			continue;
+
```
Comment (bentiss) [inline src/driver-etekcity.c]: you are in `etekcity_read_profile()`, you probably shouldn't be checking for the dirty flag.

Also, etekcity and roccat are both using the old write profile stubs, and so they do not need to check for the dirty flag as you are already checking it in `libratbag.c`

File: src/driver-gskill.c [hunk 4]
```diff
@@ -1113,6 +1113,9 @@ gskill_update_resolutions(struct ratbag_profile *profile)
 			&report->dpi_levels[report->dpi_num];
 
 		res = ratbag_profile_get_resolution(profile, i);
+		if (!res->dirty)
+			continue;
+
```
Comment (bentiss) [inline src/driver-gskill.c]: gskill starts by clearing the memory allocated with the resolutions. I am sure you are basically resetting the DPI now. gskill shouldn't check for the dirty flag.

File: src/driver-hidpp10.c [hunk 5]
```diff
@@ -574,6 +574,9 @@ hidpp10drv_commit(struct ratbag_device *device)
 
 		/* Update the current resolution in case it changed */
 		if (active_resolution) {
+			if (!resolution->dirty)
+				continue;
+
```
Comment (bentiss) [inline src/driver-hidpp10.c]: This must be OK, not entirely sure however

File: src/driver-logitech-g300.c [hunk 6]
```diff
@@ -549,6 +549,9 @@ logitech_g300_write_profile(struct ratbag_profile *profile)
 	report = &pdata->report;
 
 	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (!resolution->dirty)
+			continue;
+
```
Comment (bentiss) [inline src/driver-logitech-g300.c]: Not entirely sure if this will break or not the current resolution...

File: src/driver-roccat.c [hunk 7]
```diff
@@ -731,6 +731,9 @@ roccat_read_profile(struct ratbag_profile *profile)
 	}
 
 	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (!resolution->dirty)
+			continue;
+
```
Comment (bentiss) [inline src/driver-roccat.c]: Same than etekcity, you are in `read_profile()`

**Comment by FFY00** [comment]
Thanks for the input. I guided myself mostly from the other dirty flag checks. I will make the necessary changes when I get home.

**Comment by FFY00** [comment]
I only left the check on the steelseries driver.

**Comment by phomes** [comment]
Can you squash the patches and force push a new patch?

The sum of the two patches create a white space change in gskill.
The change in ratbag_old_write_profile should also be reverted.

**Comment by FFY00** [comment]
@phomes done.

**Comment by FFY00** [comment]
Hum, weird. Can someone restart the CI test?

**Pull request merged**